book/scheduling-and-threading.md

```diff
@@ -98,6 +102,167 @@ class TaskQueue:
+
+``` {.python expected=False}
+class Tab:
+    def __init__(self):
+        self.task_runner = TaskRunner()
+
+    def run_script(self, url, body):
+        try:
+            print("Script returned: ", self.js.run(body))
+        except dukpy.JSRuntimeError as e:
+            print("Script", url, "crashed", e)
+
+    def load(self):
+        # ...
+
+        for script in scripts:
+            # ...
+            header, body = request(script_url, url)
+            self.task_runner.schedule_task(
+                Task(self.run_script, script_url body))            
+```
+
+Now we just need to modify the main event loop to run the task runner each time,
+in case there is a script task to execute:
+
+``` {.python expected=False}
+if __name__ == "__main__":
+    # ...
+    while True:
+        while sdl2.SDL_PollEvent(ctypes.byref(event)) != 0:
+            # ...
+            browser.tabs[browser.active_tab].task_runner.run_once()
+```
+
+That's it! Now our browser will not run scripts until after `load` has completed
+and the event loop comes around again.
+
+Before continuing, let's consider why this change is interesting. It used to be
+that we had no choice but to eval scripts right away just as they were loaded.
+But now it's pretty clear that we have a lot more control over when to run
+scripts. For example, it's easy to make a change to `TaskRunner` to only run
+one script per second, or to not run them at all during page load, or when a
+tab is not the active tab. This flexibilty is quite powerful, and we can use it
+without having to dive into the guts of a `Tab` or how it loads web pages---all
+we'd have to do is implement a new `TaskRunner` heuristic.
+
+Now for the fun part I promised. Let's implement the
+[`setTimeout`][settimeout] JavaScript API, which provides a way to run
+JavaScript a given number of milliseconds from now. In terms of the JavaScript
+and Python communication, it'll use an approach with handles, similar to the
+`addEventListener` code we added in [Chapter 9](scripts.md#event-handling).
+The *new* part will be our first use of Python threads.
+
+[settimeout]: https://developer.mozilla.org/en-US/docs/Web/API/setTimeout
+
+We'll very soon have a lot more use for threads; for now all you need to know is
+that the `threading` module allows you to create threads and run code on them
+now or at a time specified in the future. To use them, start by adding an import
+for that module:
+
+``` {.python}
+import threading
+```
+
+Implement a `set_timeout` helper function that runs a callback at a specified
+time in the future. You can do that by creating a new
+[Python thread][python-thread] via the `threading.Timer` class, which takes
+two parameters: a time delta in seconds from now, and a function to call when
+that time expires.
+
+[python-thread]: https://docs.python.org/3/library/threading.html
+
+``` {.python}
+def set_timeout(func, sec):     
+    t = threading.Timer(sec, func)
+    t.start()
```

Comment:
I'm not totally clear why it's worth making a separate function for this—`threading.Timer(sec, func).start()` is no longer than calling `set_timeout`...

Reply:
The aim was:

* Introduce Python threading in baby steps, without having to get into how it's used
* Hide that behind a method that may be more familiar once you know how setTimeout is supposed to be used in JS

Perhaps that's not a good enough set of reasons..